game-app/game-core/src/main/java/games/strategy/triplea/ui/EditPanel.java

```diff
@@ -480,68 +483,95 @@ public void actionPerformed(final ActionEvent event) {
+            final PlayerChooser playerChooser =
+                new PlayerChooser(getData().getPlayerList(), getMap().getUiContext(), false);
+            final JDialog dialog =
+                playerChooser.createDialog(getTopLevelAncestor(), "Change Resources");
+            dialog.setVisible(true);
+            return Optional.ofNullable(playerChooser.getSelected());
+          }
+
+          private Optional<Resource> chooseResource() {
+            // Ignore VPS resources, since that's what the economy panel does.
+            final List<Resource> resources =
+                getData().getResourceList().getResources().stream()
+                    .filter(r -> !r.getName().equals(Constants.VPS))
+                    .collect(Collectors.toList());
+            if (resources.size() == 1) {
+              return Optional.of(resources.get(0));
+            }
+
+            final ResourceChooser chooser = new ResourceChooser(resources, getMap().getUiContext());
+            return Optional.ofNullable(
+                chooser.showDialog(getTopLevelAncestor(), "Change Resources"));
+          }
+
+          private Optional<Integer> chooseResourceValue(
+              final GamePlayer player, final Resource resource) {
+            final int oldTotal = player.getResources().getQuantity(resource.getName());
+            final JTextField totalField = new JTextField(String.valueOf(oldTotal), 4);
+            totalField.setMaximumSize(totalField.getPreferredSize());
             final int option =
                 JOptionPane.showOptionDialog(
                     getTopLevelAncestor(),
-                    new JScrollPane(pusField),
-                    "Select new number of PUs",
+                    new JScrollPane(totalField),
+                    "Select new number of " + resource.getName(),
                     JOptionPane.OK_CANCEL_OPTION,
                     JOptionPane.PLAIN_MESSAGE,
                     null,
                     null,
                     null);
             if (option != JOptionPane.OK_OPTION) {
-              cancelEditAction.actionPerformed(null);
-              return;
+              return Optional.empty();
             }
             int newTotal = oldTotal;
             try {
-              newTotal = Integer.parseInt(pusField.getText());
+              newTotal = Integer.parseInt(totalField.getText());
             } catch (final Exception e) {
               // ignore malformed input
             }
-            final String result =
-                EditPanel.this.frame.getEditDelegate().changePUs(player, newTotal);
-            if (result != null) {
-              JOptionPane.showMessageDialog(
-                  getTopLevelAncestor(),
-                  result,
-                  "Could not perform edit",
-                  JOptionPane.ERROR_MESSAGE);
-            }
-            cancelEditAction.actionPerformed(null);
+            return Optional.of(newTotal);
```

Comment:
Rather than returning the old total, why not return an optional empty? (sorry for  not looking this up myself) - Does the empty optional return value trigger a no-op or an error message?

Reply:
Actually, changed to IntTextField which eliminates the need to parse an int.